### PR TITLE
Bugfix/streamline fastify request to be generic

### PR DESCRIPTION
**Pull request by MarcoLeko**
#### Checklist

- [X] run `npm run test` and `npm run benchmark`
- [X] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [X] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

Initiated by this bug: https://github.com/fastify/fastify/issues/3775
I created a PR to fix the typings for the routeHandlers by using the `FastifyRequest` with a generic logger type.
As in the docs mentioned: https://www.fastify.io/docs/latest/Reference/Logging/#logging

> You can also supply your own logger instance. Instead of passing configuration options, pass the instance. The logger you supply must conform to the Pino interface; that is, it must have the following methods: info, error, debug, fatal, warn, trace, child.

However with the implemented type definitions it is not possible to use a custom LoggerInstance as the log property was only restricted to be of type `FastifyLoggerInstance` and no generic.

This is somewhat of an urgent bugfix from our side as it blocks our development and migration from express to fastify. This is also the reason why we set the PR against this branch as we hope to see the fix already in version 3.x. and help other developers as well.

-- Related PR for v3.x: https://github.com/fastify/fastify/pull/3782

### CODE REVIEW

File: test/types/hooks.test-d.ts [hunk 1]
```diff
@@ -123,8 +129,8 @@ server.addHook('onRegister', (instance, opts, done) => {
   expectType<void>(done(new Error()))
 })
 
-server.addHook('onReady', function (done) {
-  expectType<FastifyInstance>(this)
+server.addHook('onReady', function (instance, done: HookHandlerDoneFunction) {
+  expectType<FastifyInstance>(instance)
```
Comment (mcollina) [inline test/types/hooks.test-d.ts:133]: why this change? `this` is of type `FastifyInstance`.

File: test/types/hooks.test-d.ts [hunk 2]
```diff
@@ -202,8 +208,8 @@ server.addHook('onRegister', async (instance, opts) => {
   expectType<RegisterOptions & FastifyPluginOptions>(opts)
 })
 
-server.addHook('onReady', async function () {
-  expectType<FastifyInstance>(this)
+server.addHook('onReady', async function (instance) {
+  expectType<FastifyInstance>(instance)
```
Comment (mcollina) [inline test/types/hooks.test-d.ts:212]: Why this change?

File: test/types/hooks.test-d.ts [hunk 3]
```diff
@@ -213,11 +219,15 @@ server.addHook('onClose', async (instance) => {
...
 RawRequestDefaultExpression,
 RawReplyDefaultExpression,
-Record<string, unknown>
-> = async function (request, reply) {
+RouteGenericInterface,
+ContextConfigDefault,
+FastifySchema,
+FastifyTypeProviderDefault,
+ResolveFastifyRequestType<FastifyTypeProviderDefault, FastifySchema, RouteGenericInterface>
+> = async function (request, reply): Promise<void> {
```
Comment (mcollina) [inline test/types/hooks.test-d.ts:230]: Why this change?

File: test/types/type-provider.test-d.ts [hunk 4]
```diff
@@ -286,7 +286,7 @@ expectError(server.withTypeProvider<TypeBoxProvider>().get(
     }
   },
   async (_, res) => {
-    return false
+    res.send(false)
```
Comment (mcollina) [inline test/types/type-provider.test-d.ts:289]: why this change?

File: test/types/type-provider.test-d.ts [hunk 5]
```diff
@@ -372,7 +372,7 @@ expectError(server.withTypeProvider<JsonSchemaToTsProvider>().get(
...
```
Comment (mcollina) [inline test/types/type-provider.test-d.ts:375]: Why this change?

File: types/hooks.d.ts [hunk 6]
```diff
@@ -417,16 +458,28 @@ export interface onRegisterHookHandler<
...
+export interface onReadyHookHandler<
+  RawServer extends RawServerBase = RawServerDefault,
+  RawRequest extends RawRequestDefaultExpression<RawServer> = RawRequestDefaultExpression<RawServer>,
+  RawReply extends RawReplyDefaultExpression<RawServer> = RawReplyDefaultExpression<RawServer>,
+  Logger extends FastifyLoggerInstance = FastifyLoggerInstance,
+  TypeProvider extends FastifyTypeProvider = FastifyTypeProviderDefault,
+> {
   (
-    this: FastifyInstance,
+    instance: FastifyInstance<RawServer, RawRequest, RawReply, Logger, TypeProvider>,
```
Comment (mcollina) [inline types/hooks.d.ts:469]: I don't think this change is correct. `this` is set to the fastify isntance, and the first parameter is optional.

File: types/hooks.d.ts [hunk 7]
```diff
@@ -417,16 +458,28 @@ export interface onRegisterHookHandler<
...
+export interface onReadyAsyncHookHandler<
+  RawServer extends RawServerBase = RawServerDefault,
+  RawRequest extends RawRequestDefaultExpression<RawServer> = RawRequestDefaultExpression<RawServer>,
+  RawReply extends RawReplyDefaultExpression<RawServer> = RawReplyDefaultExpression<RawServer>,
+  Logger extends FastifyLoggerInstance = FastifyLoggerInstance,
+  TypeProvider extends FastifyTypeProvider = FastifyTypeProviderDefault,
+> {
   (
-    this: FastifyInstance,
+    instance: FastifyInstance<RawServer, RawRequest, RawReply, Logger, TypeProvider>,
```
Comment (mcollina) [inline types/hooks.d.ts:482]: Same as before.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: test/types/hooks.test-d.ts (hunk 3 above)
Comment (MarcoLeko) [reply test/types/hooks.test-d.ts:230]: Seems like when using `RawServerBase` type I get the following error: 
```ts
  test/types/hooks.test-d.ts:231:30
  ✖  231:30  Argument of type FastifyInstance<RawServerBase, IncomingMessage, ServerResponse, FastifyLoggerInstance, FastifyTypeProviderDefault> is not assignable to parameter of type FastifyInstance<Server, IncomingMessage, ServerResponse, FastifyLoggerInstance, FastifyTypeProviderDefault>.
  Types of property server are incompatible.
    Type RawServerBase is not assignable to type Server.  
```

Puzzles me a bit since I did not change anything related of the server type. Since we got an alias type for `http.Server` I switched to use now: `RawServerDefault`

Comment (mcollina) [reply test/types/hooks.test-d.ts:230]: cc @fastify/typescript

Comment (fox1t) [reply test/types/hooks.test-d.ts:230]: I've cloned this and looked into it. Actually, I think that @MarcoLeko discovered some strange bug with the `FastifyInstance` generic type. This stopped working because, in `preHandlerAsyncHookHandler`, we defined this as `FastifyInstance` without passing any generic. So @MarcoLeko passed correct generics derived from `preHandlerAsyncHookHandler,` and now the error is thrown.
current: https://github.com/fastify/fastify/blob/main/types/hooks.d.ts#L165
proposed: https://github.com/MarcoLeko/fastify/blob/bugfix/streamline-fastify-request-to-be-generic/types/hooks.d.ts#L185

@fastify/typescript any ideas?

Comment (fox1t) [reply test/types/hooks.test-d.ts:230]: Ok, never mind. I understand what is going on. `RawServerBase` is a union type, and we should not use it to test our types! We should pick one of the types of union! `RawServerDefault` should be used as the test case!

**Comment by fox1t** [comment]
What bothers me now is this PR: https://github.com/fastify/fastify/pull/3622/files#diff-bfd9cd325dab3cdde7e038db80a922d22d0f4c40949ad0d1240dcfc9b964478cL30

Why was this done in the first place? 
We should be able to pass that generics to our `FastifyInstance` to decorate `this`. Unfortunately, using just `FastifyInstance` picks default values for all generics, making it pointless passing them to `preHandlerAsyncHookHandler`.

**Comment by mcollina** [comment]
I don't know, maybe we should revert that!

Could you resolve the conflicts @MarcoLeko?

**Comment by MarcoLeko** [comment]
> I don't know, maybe we should revert that!
> 
> Could you resolve the conflicts @MarcoLeko?

Done

**Comment by fox1t** [comment]
@MarcoLeko, please ensure that every change is made in [hooks.d.ts](https://github.com/fastify/fastify/pull/3622/files#diff-bfd9cd325dab3cdde7e038db80a922d22d0f4c40949ad0d1240dcfc9b964478cL30) since your PR needs that generic is reverted. We need that generics passed to every `FastifyInstance` again.

**Comment by MarcoLeko** [comment]
> @MarcoLeko, please ensure that every change is made in [hooks.d.ts](https://github.com/fastify/fastify/pull/3622/files#diff-bfd9cd325dab3cdde7e038db80a922d22d0f4c40949ad0d1240dcfc9b964478cL30) since your PR needs that generic is reverted. We need that generics passed to every `FastifyInstance` again.

Im not sure if I misunderstood you, but you would like to revert me again the changes again for the hooks e.g.:
from:
```ts
export interface preValidationHookHandler<
...
> {
  (
    this: FastifyInstance<RawServer, RawRequest, RawReply, Logger, TypeProvider>,
    ...
  ): void;
```  

to:
```ts
export interface preValidationHookHandler<
...
> {
  (
    this: FastifyInstance,
    ...
  ): void;
```

I added the generic so we type infer the `FastifyInstance` that we have from the `this` context. I think this should be the way to go or is there a specific reason why we do not want to type infer the the instance in the hooks?

**Comment by fox1t** [comment]
OH, sorry for not being clear here! What you did is the correct thing to do, I was just asking you to check if you added `this: FastifyInstance<RawServer, RawRequest, RawReply, Logger, TypeProvider>,` to every hook instead of using that `FastifyInstance`. The "revert" part was about reverting this [PR](https://github.com/fastify/fastify/pull/3622), and it is basically what you did is already reverting that PR.

**Comment by MarcoLeko** [comment]
Ahhh okay, yeah added it to every hook :)

**Review by fox1t — APPROVED** [review]

**Review by mcollina — APPROVED** [review]
lgtm
